### DISCUSSION TITLE
[Idea] Hummingbird: Icons - make it easier to use custom icons font

### DISCUSSION THREAD

**Discussion by web-cooking-factory** [PrestaShop/PrestaShop#34107] [Ideas]
Hi people !

Themes always depends on Material Icons. As a frontend developper, I integrate custom themes based on Classic and now Hummingbird. In thoses themes, we use custom icons and so create a custom icons font.

With the actual icons markup in the theme, It means replace all occurrences in tpl. Sometimes the content is a string, sometimes it's utf-8. And there is the material icons class. For readability and independency, I think it would be better to change icons markup for something more generical and easy to customize.

Example for shopping cart icon :

Replace

`<i class="material-icons header-block__icon" aria-hidden="true">shopping_cart</i>`

By

`<i class="theme-icon theme-icon-shopping_cart header-block__icon" aria-hidden="true"></i>`

And in scss, let's say I want Awesome font but it could be any icons font:
```
.theme-icon{
   @extend .fa;

   &-shopping_cart{
      @extend .fa-shopping-cart;
   } 
}
```

Trust me, it would make my life much easier :)

EDIT: after discussion, extend is not the better way to do it, let's think about the better way but the point is markup is a problem

**Comment by dennispw** [comment]
Personally I would prefer no icon library (read: _icon font_) at all. I would much rather have inlined SVG elements.

_Icon fonts_ are convenient, easy to use and implement however the font file and the accompanying CSS files are usually _very_ large. Even more so when you think about how few icons that are actually being used. I would bet that, at most, there are 10-20 different icons used. How many icons is there in the Material Icons library? What if you decide to use Font Awesome? When writing this, the latest version of Font Awesome contains *26 233* icons.

SVG sprites can make them just as reausable as any icon font.  
(I think @Oksydan might have a proof of concept of this already if I'm not mistaken?)

Also, inline SVG are rendered instantly. As long as it has proper `height` and `width` attributes set, there is no CLS and no network requests. Icon fonts *must* wait for the font and CSS files to load before they can be displayed. Preloading, caching, etc. are valid options sure but still: those files need to be present before they can be displayed.

You also have the the `title` and `desc` elements along with the `aria-label`, `aria-labelledby` attributes to improve the accessibility.

The SVG route will most likely be more tedious to begin with but I'm positive that the tradeoffs are worth it.

**Reply by web-cooking-factory** [reply]
Yes but a svg can't be treated as text like an icon font.

I only used Awesome font as an exemple. In my case I use custom icons so just the necessary ones. No performance issue.

Trust the developpers, give them the choice, sometimes we know what we're doing ^^

**Reply by dennispw** [reply]
What do you mean by "text like an icon font" in this case?

Doesn't the `title`, `desc`, `aria-label` and `aria-labelledby` does just that? You could even throw on `aria-hidden="true"` on the icon and a `<span class="visually-hidden">Some text instead of icon font</span>`. That way you can _see_ an icon but the text will get read out by devices such as screen readers.

"Trust the developers" is one thing I would really, really like to do but over the years I've seen some pretty outlandish stuff both in modules and themes by some quite "big and trusted" developers so... :)

I'm simply saying this needs to be thought through and evaluated before throwing any icon font in to begin with. Even more so with Hummingbord so that it's done as right as possible from the get-go.

**Reply by web-cooking-factory** [reply]
I mean apply CSS font-size, color, ...

What I meant by trust us is : please don't close doors. SVG doesn't fit my needs. We should have the choice.

**Reply by dennispw** [reply]
Oh absolutely, I'm all for having options. I'm not trying to convince you that _you_ should use SVG in whatever project you're working on, not at all! If you need or want to use icons fonts, go for it!

You did open this up for the _Hummingbird_ theme so I'm weighing in my opinion that the theme should not be tied down to icon fonts. If the theme is shipped with an icon font - make it easy to remove and/or alter it to fit our needs. _Personally_ I am still voting for the switch to SVG.

<sub>By the way you could match the SVG to any text using `em`, `currentColor` and so on if you really need to ;)</sub>

**Reply by Oksydan** [reply]
Hi @dennispw,

I will try to create a POC tonight.

**Reply by Oksydan** [reply]
POC added https://github.com/PrestaShop/hummingbird/pull/551

**Comment by matks** [comment]
Hi everyone in order to provide more visibility for the community I moved this Discussion into PrestaShop/PrestaShop Discussions. The Hummingbird repository might not attract enough attention as suggested by @Oksydan
